### PR TITLE
Fixes #57 Remove server load balancing and use single replication acceptor

### LINKED ISSUE #57: Few adjustments on the Artemis Test
I've filing the 2 things I believe should be fixed on the ActiveMQ Artemis test:

- `netty-connector` and `netty-acceptor` are the same and clients are connecting to `netty-acceptor`: I suggest to keep `netty-acceptor` as the acceptor responsible to deal with clients connections and define a new acceptor (let's call it `cluster-acceptor`, configured with `protocols=CORE` and  `remotingThreads=1`) that would handle the connectivity with the other broker (could be either `master` or `slave`), and `netty-connector` should be use this `cluster-acceptor` address instead
- as said on https://github.com/softwaremill/mqperf/pull/47#issuecomment-778062529 the test shouldn't use the other live brokers to distribute the traffic, given that Artemis won't guarantee automatic balanced resource partitions

The former is quite important to guarantee good results given that replicating is mandatory to return backup responses to clients and having a separate connection with its own thread guarantee to be able to response promptly.

### PR DESCRIPTION
**Pull request by franz1981**
Fixes #57

### CODE REVIEW

**Comment by franz1981** [comment]
Some of these changes can already be effective, but https://github.com/apache/activemq-artemis/pull/3392 is a game-changer in term of the replication broker performance, hence only `master` or next release will show the most of the benefits of these changes.

**Comment by franz1981** [comment]
@adamw Let me know if I can help to demystify these somehow ^^

**Comment by franz1981** [comment]
@adamw FYI: just running this changes vs 2.17 (that's not including the improvements I've mentioned on https://github.com/softwaremill/mqperf/pull/58#issuecomment-803857883), I'm getting:

`baseline` from https://softwaremill.com/mqperf/#activemq-artemis

image

with this PR:
new_25_4_8_sender

TLDR

~53 K msg/sec vs ~87 K msg/sec producer rate

**Comment by franz1981** [comment]
With this changes on `artemis` master this PR is now getting:

image

on steady state, but I see we've more latency outlier; I don't know how the public blog post has measured latencies during JVM warmup

**Comment by franz1981** [comment]
Just a note for the blog maintainers: I've run the same benchmark of https://github.com/softwaremill/mqperf/pull/58#issuecomment-804772617 4 times without getting any latency outlier, see:

warming_up_2

NOTE: the rates are different because I've collected startup and shutdown too into the snapshot.

That means that mqperf tests *should* account a warming up phase for 2 reasons:
- "cold heap" (you can set `-XX:+AlwaysPreTouch` or just run the test several times to let most of the needed heap memory being "touched")
- Java compilation quirks

**Comment by adamw** [comment]
Thanks a lot for your reports and the updated config :) We'll try to refresh the blog in some near time.

When reporting measurments we always discard the initial warm-up (and the cool-down) sections, taking only the "middle" where all producers/consumers report a steady rate of message processing.

**Comment by franz1981** [comment]
@adamw 

>When reporting measurments we always discard the initial warm-up (and the cool-down) sections, taking only the "middle" where all producers/consumers report a steady rate of message processing.

Great, but probably that's not "enough", especially for JVM applications that heavily relies on heap and JIT compilations, see my comment about `-XX:+AlwaysPreTouch` too (and consider that the AMQ broker and maybe other brokers will heavily use heap and need it to be correctly allocated and "touched" upfront).

Kafka IIRC isn't heavily relying on the heap or JIT compilation given that is mostly I/O bounded, while Artemis, yep: I start to get the "correct" latencies after run the mqperf benchmark 3 or 4 times, given that we've set 48 GB of heap (!!).

**Comment by franz1981** [comment]
The last change is disabling broker-side load balancing, leaving just clients to load balance among broker nodes: as said in another context, broker load balancing cause distribution of messages among cluster nodes because it assumes consumers aren't evenly distributed, adding a coherency traffic that prevent the brokers to correctly scale.

**Comment by franz1981** [comment]
I've run the bench vs these changes again, playing a bit with the number of nodes:

image

And I'm still not totally convinced the `mqperf` is comparing brokers apple 2 apple in term of semantic (that's ok given we're talking about very different products). 
On the Artemis broker an anycast queue is a resource shared among producers/consumers and especially consumers dequeue it  in a single-thread fashion ie adding consumers for the same queue won't improve performance, but the opposite(!!!) that's what's showing the above picture, if compared with https://user-images.githubusercontent.com/13125299/112326121-8b0fc300-8cb4-11eb-8ac1-413b6d460fca.png

The above picture is indeed comparing 3 different tests:
- 25 threads, 4 producers, 4 consumers
- 25 threads, 6 producers, 4 consumers
- 25 threads, 8 producers, 4 consumers

In all cases the throughput is increased without any huge impact on latencies.
There is a breaking point by which the single-threaded processing of consumers just start to get negative scaling while adding consumer nodes: considering that clients are not interested on total order (per-node), probably using a single queue isn't the right tool to scale (considering the JMS semantic) and instead, using topics or just using different set of queues (that's a more verbose version of using partitions on Kafka, really) it's the way to go to scale with the number of consumers.

**Comment by franz1981** [comment]
I see that by using FQQN queues on Artemis, probably we can achieve the right semantic:
image

The concept of addresses and anycast queues of Artemis is near to Kafka partitions (within a consumer group) and should allow consumers to scale: I'll experiment with this a bit more in the next days. 
I don't know how to configure mqperf receivers to correctly match a specific queue 1:1 so I am opened to suggestions.

**Comment by adamw** [comment]
The receivers can be configured using `TestConfig`, where you can access the default settings as well as arbitrary ones via `TestConfig.mqConfig`. However the assumption is that when a broker leaves / joins the cluster, the recievers shouldn't need reconfiguration, so the config should be quite static

**Comment by franz1981** [comment]
Thanks @bturos for the merge!
I suggest to run the next Artemis bench round using artemis master instead of 2.16, because https://github.com/apache/activemq-artemis/pull/3392 isn't yet in any release if not on master, but is stable from some time on our CI. 
Due to summer we're awaiting a bit before the next release, but would be great to get confirmation in your env if the improvement is as good as it seems in our perf tests :)

Said that, this benchmark is not yet right with the address model of artemis (see https://github.com/softwaremill/mqperf/pull/58#issuecomment-810589075 comment), because it basically force each broker to deal with a huge amount of producers/consumers against a single queue, making it a huge single threaded beast: in this case having less cores but more frequency is key to make it scales.
In addition, it forces total (per broker) ordering among producers/consumers while the other brokers give up on total ordering to improve scalability.
As you can see, it's a very different scenario with obvious performance trade-offs.

We haven't yet provided a feature that allow exclusive but re-balanced assignments of queue partitions to group of consumers (if not the FQQN mechanism I've already mentioned, but is far from being "automatic") and I would like this one to be clear while commenting benchmarks results: in the current scenario, despite the increased network (and disk) bandwidth requirements, probably a solution based on topics (vs queues) where each consumer filter useless/unwanted messages (based on some ad-hoc message property) will perform better, because will use different delivering threads for each subscriber, instead of a single one.

In addition, probably a JMS 2 implementation using async producers can help improving client machine thread usage too, wdyt?

**Comment by bturos** [comment]
Hi @franz1981 - first of all, thanks for active participation on this project, this really stands out.

We had an internal discussion on the topics mentioned above. In the order of appearance:
* testing against master branch - sadly, that's not possible. We test only against public releases.
  Having said that, we'll be happy to get back to this in the future, once the changes mentioned
  above are part of official Artemis release.

* changing the topology - this tests model assumes that there is a single queue with a number of producers and consumers. We don't make a difference between any of the consumers. Consuming different queues/partitions means introducing some kind of state management which increases complexity of the implementation. Looking forward to any suggestions on how that could be achieved on the broker/client API side at the moment.

* JMS2 / async producers - out of scope for this round

**Comment by franz1981** [comment]
hi @bturos 

> Looking forward to any suggestions on how that could be achieved on the broker/client API side at the moment.

I'm talking about this with the team to see if it turns to be some nice feature that can ease the life for users too :) 
The address model of Artemis is pretty flexible, but as you said, we need some kind of automation to make users to 
use it for use cases like this. 
Said that, my point was meant to underline during an eventual analysis of results that the mqperf tests are single-core with the current feature-set of the broker, hence increasing consumers/producers hit the single-threaded execution wall that prevent scalability to fly. 

Re 
> Having said that, we'll be happy to get back to this in the future, once the changes mentioned
above are part of official Artemis release.

When the next round of tests/benchmark results are planned to be run?

**Comment by bturos** [comment]
Hi @franz1981 

Artemis scalability -> sure, that's on our list of things we need to add in the article once the current round of tests is done.

Next round -> no fixed date on that. Once we know there are some significant improvements in the tested products AND we have free CPU cycles to tackle the topic, we'll get on it.
